File: 472-concatenated-words/472-concatenated-words.py
class Solution:
    def findAllConcatenatedWordsInADict(self, words: List[str]) -> List[str]:
        def dfs(word, wordSet):
            for i in range(1, len(word)):
                prefix = word[:i]
                postfix = word[i:]
                if prefix in wordSet and postfix in wordSet:
                    return True
                if prefix in wordSet and dfs(postfix, wordSet):
                    return True
                if postfix in wordSet and dfs(prefix, wordSet):
                    return True
            return False
        
        d = set(words)
        res = []
        for word in words:
            if dfs(word, d):
                res.append(word)
        return res
        

File: 472-concatenated-words/472-concatenated-words.java
class Solution {
    public List<String> findAllConcatenatedWordsInADict(String[] words) {
        List<String> ans = new ArrayList<>();
        Arrays.sort(words, (a, b) -> a.length() - b.length());
        Set<String> preWords = new HashSet<>();
        
        for (String s : words) {
            if (canForm(s, preWords)) ans.add(s);
            preWords.add(s);
        }
        return ans;
    }
    
    private boolean canForm(String s, Set<String> preWords) {
        if (preWords.isEmpty()) return false;
        
        boolean[] memo = new boolean[s.length() + 1];
        memo[0] = true;
        for (int right = 1; right <= s.length(); right++) {
            for (int left = 0; left < right; left++) {
                if (memo[left] == true && preWords.contains(s.substring(left, right))) memo[right] = true;
            }
        }
        return memo[s.length()];
    }
}